File: src/csharp/Tensor.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Diagnostics;

namespace Microsoft.ML.OnnxRuntimeGenAI
{
    /// <summary>
    /// Element types that correspond to OnnxRuntime supported element types.
    /// </summary>
    /// The values in this enum must match ONNX Runtime's type values
    public enum ElementType : long
    {
        undefined,
        float32,
        uint8,
        int8,
        uint16,
        int16,
        int32,
        int64,
        string_t,
        bool_t,
        float16,
        float64,
        uint32,
        uint64,
    }

    ///<summary>
    /// Currently wraps an ORT Tensor.
    ///</summary>
    public class Tensor : IDisposable
    {
        private IntPtr _tensorHandle;
        private bool _disposed = false;

        /// <summary>
        /// Constructs a Tensor with the given data pointer, shape and element type.
        /// </summary>
        /// <param name="data">The data pointer.</param>
        /// <param name="shape">The shape of the Tensor.</param>
        /// <param name="type">The type of elements in the Tensor.</param>
        /// <exception cref="OnnxRuntimeGenAIException">
        /// Thrown when the call to the GenAI native API fails.
        /// </exception>
        public Tensor(IntPtr data, Int64[] shape, ElementType type)
        {
            Result.VerifySuccess(NativeMethods.OgaCreateTensorFromBuffer(data, shape, (UIntPtr)shape.Length, type, out _tensorHandle));
        }

        internal Tensor(IntPtr tensorHandle)
        {
            Debug.Assert(tensorHandle != IntPtr.Zero);
            _tensorHandle = tensorHandle;
            _disposed = false;
        }

        internal IntPtr Handle { get { return _tensorHandle; } }

        ~Tensor()
        {
            Dispose(false);
        }

        /// <summary>
        /// Get the element type.
        /// </summary>
        /// <returns>
        /// The element type.
        /// </returns>
        /// <exception cref="OnnxRuntimeGenAIException">
        /// Thrown when the call to the GenAI native API fails.
        /// </exception>
        public ElementType Type()
        {
            Result.VerifySuccess(NativeMethods.OgaTensorGetType(_tensorHandle, out ElementType element_type));
            return element_type;
        }

        /// <summary>
        /// Get the tensor shape.
        /// </summary>
        /// <returns>
        /// The shape.
        /// </returns>
        /// <exception cref="OnnxRuntimeGenAIException">
        /// Thrown when the call to the GenAI native API fails.
        /// </exception>
        public Int64[] Shape()
        {
            Result.VerifySuccess(NativeMethods.OgaTensorGetShapeRank(_tensorHandle, out UIntPtr size));
            Int64[] shape = new Int64[size.ToUInt64()];
            Result.VerifySuccess(NativeMethods.OgaTensorGetShape(_tensorHandle, shape, size));
            return shape;
        }

        /// <summary>
        /// Computes number of elements in the tensor
        /// given the shape
        /// </summary>
        /// <param name="shape">shape</param>
        /// <returns>product of dimensions</returns>
        public static Int64 ElementsFromShape(Int64[] shape)
        {
            Int64 size = 1;
            foreach (Int64 dim in shape)
            {
                size *= dim;
            }
            return size;
        }

        /// <summary>
        /// Computes and returns number of elements in the tensor
        /// </summary>
        /// <returns>The number of elements.</returns>
        public Int64 NumElements()
        {
            return ElementsFromShape(Shape());
        }

        /// <summary>
        /// Return a ReadOnlySpan to tensor data
        /// no type checks are made
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <returns>read only span</returns>
        /// <exception cref="OnnxRuntimeGenAIException">
        /// Thrown when the call to the GenAI native API fails.
        /// </exception>
        public ReadOnlySpan<T> GetData<T>()
        {
            var elements = NumElements();
            Result.VerifySuccess(NativeMethods.OgaTensorGetData(Handle, out IntPtr data));
            unsafe
            {
                return new ReadOnlySpan<T>(data.ToPointer(), (int)elements);
            }
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        protected virtual void Dispose(bool disposing)
        {
            if (_disposed)
            {
                return;
            }
            NativeMethods.OgaDestroyTensor(_tensorHandle);
            _tensorHandle = IntPtr.Zero;
            _disposed = true;
        }
    }

    /// <summary>
    /// This class is a list of tensors with names that match up with model input names.
    /// </summary>
    public class NamedTensors : IDisposable
    {
        private IntPtr _namedTensorsHandle;
        private bool _disposed = false;

        internal NamedTensors(IntPtr namedTensorsHandle)
        {
            _namedTensorsHandle = namedTensorsHandle;
        }

        internal IntPtr Handle { get { return _namedTensorsHandle; } }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (_disposed)
            {
                return;
            }
            NativeMethods.OgaDestroyNamedTensors(_namedTensorsHandle);
            _namedTensorsHandle = IntPtr.Zero;
            _disposed = true;
        }
    }
}

File: src/java/src/main/java/ai/onnxruntime/genai/Tensor.java
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
package ai.onnxruntime.genai;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/** Currently wraps an ORT Tensor. */
public final class Tensor implements AutoCloseable {
  private long nativeHandle = 0;
  private final ElementType elementType;
  private final long[] shape;

  // Buffer that owns the Tensor data.
  private ByteBuffer dataBuffer = null;

  // The values in this enum must match ONNX values
  // https://github.com/onnx/onnx/blob/159fa47b7c4d40e6d9740fcf14c36fff1d11ccd8/onnx/onnx.proto#L499-L544
  /** Element types that correspond to OnnxRuntime supported element types. */
  public enum ElementType {
    /** Undefined element type. */
    undefined,
    /** 32-bit IEEE 754 floating point. */
    float32,
    /** 8-bit unsigned integer. */
    uint8,
    /** 8-bit signed integer. */
    int8,
    /** 16-bit unsigned integer. */
    uint16,
    /** 16-bit signed integer. */
    int16,
    /** 32-bit signed integer. */
    int32,
    /** 64-bit signed integer. */
    int64,
    /** UTF-8 encoded string. */
    string,
    /** Boolean value (true or false). */
    bool,
    /** 16-bit IEEE 754 floating point (half precision). */
    float16,
    /** 64-bit IEEE 754 floating point (double precision). */
    float64,
    /** 32-bit unsigned integer. */
    uint32,
    /** 64-bit unsigned integer. */
    uint64,
  }

  /**
   * Constructs a Tensor with the given data, shape and element type.
   *
   * @param data The data for the Tensor. Must be a direct ByteBuffer with native byte order.
   * @param shape The shape of the Tensor.
   * @param elementType The type of elements in the Tensor.
   * @throws GenAIException If the call to the GenAI native API fails.
   */
  public Tensor(ByteBuffer data, long[] shape, ElementType elementType) throws GenAIException {
    if (data == null || shape == null || elementType == ElementType.undefined) {
      throw new GenAIException(
          "Invalid input. data and shape must be provided, and elementType must not be undefined.");
    }

    // for now require the buffer to be direct.
    // we could support non-direct but need to do an allocate and copy here.
    if (!data.isDirect()) {
      throw new GenAIException(
          "Tensor data must be direct. Allocate with ByteBuffer.allocateDirect");
    }

    // for now, require native byte order as the bytes will be used directly.
    if (data.order() != ByteOrder.nativeOrder()) {
      throw new GenAIException("Tensor data must have native byte order.");
    }

    this.elementType = elementType;
    this.shape = shape;
    this.dataBuffer = data; // save a reference so the owning buffer will stay around.

    nativeHandle = createTensor(data, shape, elementType.ordinal());
  }

  /**
   * Construct a Tensor from native handle.
   *
   * @param handle The native tensor handle.
   */
  Tensor(long handle) {
    nativeHandle = handle;
    elementType = ElementType.values()[getTensorType(handle)];
    shape = getTensorShape(handle);
  }

  /**
   * Get the element type.
   *
   * @return The element type.
   */
  public ElementType getType() {
    return this.elementType;
  }

  /**
   * Get the tensor shape.
   *
   * @return The tensor type.
   */
  public long[] getShape() {
    return this.shape;
  }

  @Override
  public void close() {
    if (nativeHandle != 0) {
      destroyTensor(nativeHandle);
      nativeHandle = 0;
    }
  }

  long nativeHandle() {
    return nativeHandle;
  }

  static {
    try {
      GenAI.init();
    } catch (Exception e) {
      throw new RuntimeException("Failed to load onnxruntime-genai native libraries", e);
    }
  }

  private native long createTensor(ByteBuffer data, long[] shape, int elementType)
      throws GenAIException;

  private native void destroyTensor(long tensorHandle);

  private native int getTensorType(long tensorHandle);

  private native long[] getTensorShape(long tensorHandle);
}
